### PR TITLE
#167971251 Implement user-data validation using express-validator

### PR DESCRIPTION
**Pull request by Oliver-ke**
#### Description of Task to be completed?
- Add express-validator to project dependencies
- Add validate middleware
- Add userRegister validation rules
- Add userLogin validation rules

 #### How should this be manually tested?
- Run `npm install` to install dependencies
- Start  the server with `npm run start:dev`
- Using postman make a POST request with JSON data to `localhost:3000/api/v1/users`

 #### Any background context you want to provide?
None
 #### What are the relevant pivotal tracker stories?
#[167971251](https://www.pivotaltracker.com/story/show/167971251)
 #### Screenshots (if appropriate)
None
 #### Questions
None

### CODE REVIEW

**Review by feobaby — CHANGES_REQUESTED** [review]
Goo job K, first things first, change the PR title to  `Implement user-data validation using express-validator`

**Review by feobaby — DISMISSED** [review]
yes, so I have tested it using postman and also the unit tests. everything works fine. Nice work.

File: src/index.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import express from 'express';
 import cors from 'cors';
+import validate from './middlewares/validate';
```
Comment (oyewoas) [inline src/index.js]: Well done, I think you can just import this to **index.js** in the middleware folder and export it from there, this will be a better way of exporting middleware functions because we can be sure that all middleware is being exported from index.js. This helps in a case where we have more than one middleware apart from **validate.js** in the middleware folder, we can just refer to that middleware by importing the index.js in the middleware folder then use it as needed in the codebase. or what do you think @Oliver-ke, let me know if you understand what i am suggesting..

Comment (feobaby) [reply src/index.js]: @oyewoas I think he did ex2what you just explained src/index.  That is the entry point file now?

File: src/test/validate.test.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+chai.should();
+
+describe('validate middleware test', () => {
+  it('Should return error for invalid post data', (done) => {
+    const dummyUser = {
+      firstName: 'john',
+      lastName: 'doe',
+      password: 'password',
+      email: 'invalidMail'
+    };
```
Comment (oyewoas) [inline src/test/validate.test.js:27]: Welldone, i think firstName is meant to be first_name and lastName is meant to be last_name since that is what will be used in the user model. I got an error on postman when I used first_name and last_name as properties...

Comment (feobaby) [reply src/test/validate.test.js:27]: @oyewoas Well the TTL said he does not want to see any form of snake casing...

File: src/index.js (hunk 1 above)
Comment (oyewoas) [reply src/index.js]: I was actually talking about middleware/index.js, in a case where we have **validate.js**, **somethingelse.js**, we can import them into **middleware/index.js**, and export the functions in each from there, so that if we need the middleware for other parts of the codebase we just import from middleware/index.js and not have to be calling middleware/validate.js, and middleware/somethingelse.js.. do you get me @funmi5

Comment (feobaby) [reply src/index.js]: @oyewoas hm, I totally prefer the way he did his, also you have a point but I still really don't get you.

File: src/test/validate.test.js (hunk 2 above)
Comment (oyewoas) [reply src/test/validate.test.js:27]: > @oyewoas Well the TTL said he does not want to see any form of snake casing...

ok, i prefer that too

File: src/index.js (hunk 1 above)
Comment (oyewoas) [reply src/index.js]: > @oyewoas I think he did ex2what you just explained src/index. That is the entry point file now?

I was actually talking about middleware/index.js, in a case where we have validate.js, somethingelse.js, we can import them into middleware/index.js, and export the functions in each from there, so that if we need the middleware for other parts of the codebase we just import from middleware/index.js and not have to be calling middleware/validate.js, and middleware/somethingelse.js.. do you get me @funmi5

Comment (oyewoas) [reply src/index.js]: > @oyewoas hm, I totally prefer the way he did his, also you have a point but I still really don't get you.

its not a pressing thing. and it works all good..

**Review by oyewoas — DISMISSED** [review]
well done

File: src/index.js (hunk 1 above)
Comment (eokwukwe) [reply src/index.js]: The concern raised by @oyewoas has not been addressed yet and it is a valid based on our directory structure of the project. When are you planning to implement that?

File: src/index.js [hunk 3]
```diff
@@ -14,6 +15,11 @@ const isProduction = process.env.NODE_ENV === 'production';
 // Test route
 app.get('/', (req, res) => res.json({ status: res.statusCode, message: 'Welcome to Wolfsbane' }));
 
+// validation test route
+app.post('/user', validate('userRegister'), (req, res) => {
+  res.status(201).json({ message: 'success' });
+});
```
Comment (eokwukwe) [inline src/index.js]: Why is this in the place? I suppose we have a directory for declaring routes.

Comment (Oliver-ke) [reply src/index.js]: Ok @eokwukwe, I was just using it for testing, I'll move to the route folder then.

File: src/routes/index.js [hunk 4]
```diff
@@ -1,7 +1,6 @@
-import express from 'express';
 
-const router = express.Router();
+import usersRoute from './api/users';
 
-router.use('/api', require('./api'));
```
Comment (oyewoas) [inline src/routes/index.js:5]: i don't think we need to remove this line.... @Oliver-ke , we need express router here.. or what do you think ??

File: src/routes/index.js [hunk 5]
```diff
@@ -1,7 +1,6 @@
...
-
-module.exports = router;
```
Comment (oyewoas) [inline src/routes/index.js:7]: i don't think  we need to remove this line too... overall well done bro

**Review by oyewoas — CHANGES_REQUESTED** [review]

File: src/routes/index.js (hunk 4 above)
Comment (feobaby) [reply src/routes/index.js:5]: @oyewoas it can be modified with time, i think he  just went plainly with **_'api/v1/users'_**

**Review by feobaby — COMMENTED** [review]
Did you eventually write the sinon mockup?
Also, was this how he specified the error message to be:
```
{
  "errors": {
    "firstName": "firstName is invalid, firstName should be an alphabetical character, greater then 1 and less than 250 characters, should not include numbers and cannot be left empty",
    "lastName": "lastName is invalid, lastName should be an alphabetical character, greater then 1 and less than 250 characters, should not include numbers and cannot be left empty",
    "email": "Please provide a valid email"
  }
}
```

File: src/routes/index.js (hunk 4 above)
Comment (oyewoas) [reply src/routes/index.js:5]: > @oyewoas it can be modified with time, i think he just went plainly with **_'api/v1/users'_**

@funmi5 i get you, i understand what he did, I think he could have still gotten the same result without modifying that file since that convention will be used for endpoints creation.

Comment (Oliver-ke) [reply src/routes/index.js:5]: @oyewoas , bro this was really needed.. The file structure became complicated, please take a second look at it

File: src/routes/index.js (hunk 5 above)
Comment (Oliver-ke) [reply src/routes/index.js:7]: @oyewoas  we're writing es6 and we need to use es6 standard, suggest to look more into this.
All I did there was trying to organize the file structure and imports

Comment (oyewoas) [reply src/routes/index.js:7]: > @oyewoas we're writing es6 and we need to use es6 standard, suggest to look more into this.
> All I did there was trying to organize the file structure and imports

yes thats nice, you could look into example  in **routes/index.js** you can have  @Oliver-ke 
```javascript
import express from 'express';
import api from './api';

const router = express.Router();

router.use('/api', api);

export default router;
```

**Review by oyewoas — DISMISSED** [review]

**Review by uthdev — DISMISSED** [review]
Nice one @Oliver-ke, Tested and working fine. I think the error sent for `firstName` and `lastName` are kind of too long but I understand that this is just a dummy route to test the implementation and everyone would have to write their own rules for whatever they want to validate. So I believe the person working one the user registration would define rules and concise error messages.

File: src/validation/validators/rules.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { check } from 'express-validator';
+
+// add validation rules here.
+export const userRegister = [
+  check('firstName').matches(/^[\w'\-,.][^0-9_!¡?÷?¿/\\+=@#$%ˆ&*(){}|~<>;:[\]]{1,250}$/).withMessage('firstName is invalid, firstName should be an alphabetical character, greater then 1 and less than 250 characters, should not include numbers and cannot be left empty').trim(),
+  check('lastName').matches(/^[\w'\-,.][^0-9_!¡?÷?¿/\\+=@#$%ˆ&*(){}|~<>;:[\]]{1,250}$/).withMessage('lastName is invalid, lastName should be an alphabetical character, greater then 1 and less than 250 characters, should not include numbers and cannot be left empty').trim(),
```
Comment (eokwukwe) [inline src/validation/validators/rules.js]: @Oliver-ke  include an explanation of the regex

Comment (Oliver-ke) [reply src/validation/validators/rules.js]: Ok @eokwukwe, I've done that, thanks for the feedback

**Review by feobaby — DISMISSED** [review]
Great work here.

**Review by darasimiolaifa — DISMISSED** [review]
Well done here Bro. Works fine.

**Review by feobaby — DISMISSED** [review]

**Review by sa-ma — DISMISSED** [review]

**Review by feobaby — DISMISSED** [review]

**Review by darasimiolaifa — DISMISSED** [review]
Well done. Good work.

File: src/middlewares/index.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+
+// eslint-disable-next-line
```
Comment (eokwukwe) [inline src/middlewares/index.js]: Why is this here? I thought we have an `eslint` configuration file where our rules are declared

File: src/validation/validators/rules.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+this prevent us from matching numbers, $@*()+= etc
+the last {1,250} is a quantifier specifying that the character been matched
+should be > 1 and <= 251
+
+the ^ and $ runs the match from the beginning and end of the string
+*/
+export const userRegister = [
+  check('firstName')
+    .matches(/^[\w'\-,.][^0-9_!¡?÷?¿/\\+=@#$%ˆ&*(){}|~<>;:[\]]{1,250}$/)
+    .withMessage(
```
Comment (eokwukwe) [inline src/validation/validators/rules.js]: `[\w'\-,.] => any "alphanumeric and underscores", any "'-,."`. The only special characters that may be found in a name is `' or -`. Also, I have not seen where a name is made up of `alphanumeric` characters or `underscores`. 

I know you can have a name like `Ma'John` or `John-Sleg`. However, I have never seen a name like `Ma4John` or `Ma_John`. Hence the regex needs improvement.

Since the same regex is used multiple times, you may need to assign it to a variable and then use it when needed.

File: src/validation/validators/rules.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      'lastName is invalid, lastName should be an alphabetical character, greater then 1 and less than 250 characters, should not include numbers and cannot be left empty'
+    )
+    .trim(),
+  check('password', 'password should be at least 6 characters').isLength({ min: 6 }),
+  check('email', 'Please provide a valid email')
+    .isEmail()
+    .isLength({ min: 3, max: 250 })
+    .trim()
+];
+
```
Comment (eokwukwe) [inline src/validation/validators/rules.js:35]: The error messages are too long. Please look for a shorter summary for them.

File: src/index.js [hunk 10]
```diff
@@ -10,6 +13,7 @@ app.use(express.urlencoded({ extended: false }));
 app.use(express.json());
 
 const isProduction = process.env.NODE_ENV === 'production';
+app.use('/api/v1/users', usersRoute);
```
Comment (eokwukwe) [inline src/index.js]: Why is this here? This supposed to be handled in the route directory. Refer to @oyewoas's task to refactor this routing and put the logic where it belongs

File: src/routes/index.js [hunk 11]
```diff
@@ -1,7 +1,8 @@
 import express from 'express';
+import api from './api';
 
 const router = express.Router();
 
-router.use('/api', require('./api'));
+router.use('/api', api);
```
Comment (eokwukwe) [inline src/routes/index.js]: Make this `/api/v1`

File: src/test/unit-test/validate.test.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
+/* eslint-disable no-unused-expressions */
+import { expect } from 'chai';
+import middlewares from '../../middlewares';
+
+const { validate } = middlewares;
+
+describe('Validate middleware test suite', () => {
+  let mw;
```
Comment (eokwukwe) [inline src/test/unit-test/validate.test.js]: Use a proper name for variables

Comment (Oliver-ke) [reply src/test/unit-test/validate.test.js]: Ok @eokwukwe , resolved

File: src/routes/index.js (hunk 11 above)
Comment (Oliver-ke) [reply src/routes/index.js]: Ok, resolved

**Review by oyewoas — APPROVED** [review]
nice @Oliver-ke
